File: app/models/cve.py
"""Pydantic data models for KnowCVE — hybrid refactor final version."""

from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


# ── Raw NVD data ──────────────────────────────────────────────────────────────

class RawCVE(BaseModel):
    """CVE record parsed directly from the NVD API v2 response."""

    cve_id: str = Field(..., description="e.g. CVE-2024-12345")
    description: str = ""
    cvss_score: float = Field(default=0.0, ge=0.0, le=10.0)
    cvss_vector: str = ""
    cvss_version: str = ""
    published_date: Optional[datetime] = None
    last_modified: Optional[datetime] = None
    references: list[str] = Field(default_factory=list)
    weaknesses: list[str] = Field(default_factory=list)


# ── Enrichment ────────────────────────────────────────────────────────────────

class EnrichmentData(BaseModel):
    """Aggregated enrichment from CISA KEV, EPSS, GitHub advisories, and GreyNoise."""

    in_kev: bool = False
    epss_score: float = Field(default=0.0, ge=0.0, le=1.0)
    epss_percentile: float = Field(default=0.0, ge=0.0, le=1.0)
    poc_urls: list[str] = Field(default_factory=list)
    has_poc: bool = False
    greynoise_scanner_count: int = Field(default=0, ge=0)
    is_being_scanned: bool = False

    # Trend tracking (populated by trending_refresh_job)
    previous_epss_score: float = 0.0
    previous_scanner_count: int = 0
    epss_trend: str = "stable"       # "rising", "falling", "stable"
    scanner_trend: str = "stable"    # "rising", "falling", "stable", "new"

    # Nuclei template detection (Phase 4)
    has_nuclei_template: bool = False
    nuclei_template_url: Optional[str] = None


# ── ATT&CK Technique (Phase 5.5) ─────────────────────────────────────────────

class AttackTechnique(BaseModel):
    """A single MITRE ATT&CK technique in a CVE-specific kill chain."""

    technique_id: str = ""          # "T1190"
    technique_name: str = ""        # "Exploit Public-Facing Application"
    tactic: str = ""                # "Initial Access"
    tactic_phase: int = 1           # 1-12 ordering per ATT&CK tactic order
    description: str = ""           # CVE-specific application of this technique
    is_pivot: bool = False          # True if this is the key exploitation step


# ── AI Explanation ────────────────────────────────────────────────────────────

class AIExplanation(BaseModel):
    """AI-generated threat intelligence explanation of a CVE."""

    summary: str = ""
    technical_detail: str = ""
    impact: str = ""
    remediation: str = ""
    tags: list[str] = Field(default_factory=list)
    affected_tech: list[str] = Field(default_factory=list)
    mitre_techniques: list[dict] = Field(default_factory=list)
    # Each dict: {"technique_id": "T1190", "technique_name": "...", "tactic": "...", "url": "..."}

    # Phase 5.5 — AI Depth Upgrade fields
    vulnerability_class_analysis: Optional[str] = None
    adversarial_context: Optional[str] = None
    exploit_narrative: Optional[str] = None
    attack_techniques: Optional[list[AttackTechnique]] = None


# ── Processed CVE ─────────────────────────────────────────────────────────────

class ProcessedCVE(BaseModel):
    """Fully enriched + scored + explained CVE ready for API / Telegram."""

    cve_id: str
    description: str = ""
    cvss_score: float = 0.0
    cvss_vector: str = ""
    cvss_version: str = ""
    published_date: Optional[datetime] = None
    last_modified: Optional[datetime] = None
    references: list[str] = Field(default_factory=list)
    weaknesses: list[str] = Field(default_factory=list)

    enrichment: EnrichmentData = Field(default_factory=EnrichmentData)
    ai_explanation: Optional[AIExplanation] = None
    priority_score: int = Field(default=0, ge=0, le=100)
    priority_label: str = "LOW"
    processed_at: datetime = Field(default_factory=datetime.utcnow)

    # Dynamic scoring & categorisation (new)
    categories: list[str] = Field(default_factory=list)
    # e.g. ["ACTIVELY_EXPLOITED", "TRENDING", "JUST_DROPPED", "SUPPLY_CHAIN"]
    dynamic_score: int = 0
    # the momentum/dynamic portion of the score
    last_rescored_at: Optional[datetime] = None
    # when this CVE was last re-evaluated

    # Advisory source tracking (supply chain detection)
    ghsa_id: Optional[str] = None
    ecosystem: Optional[str] = None          # "npm", "PyPI", etc.
    is_malware_advisory: bool = False        # True for supply chain malware
    affected_packages: list[dict] = Field(default_factory=list)

    # KRS (KnowCVE Risk Score) aliases — Phase 4
    @property
    def krs_score(self) -> int:
        return self.priority_score

    @property
    def krs_label(self) -> str:
        return self.priority_label

    model_config = {"arbitrary_types_allowed": True}


# ── Telegram Subscriptions ───────────────────────────────────────────────────

class UserSubscription(BaseModel):
    """A Telegram user's alert subscription."""

    chat_id: int
    subscribed_at: datetime = Field(default_factory=datetime.utcnow)
    min_priority: int = Field(default=25, ge=0, le=100, description="Minimum priority score to receive alerts")
    alert_mode: str = Field(default="important", description="continuous|important|critical")
    paused: bool = Field(default=False, description="When True, no alerts sent until resumed")


# ── API Response Models ──────────────────────────────────────────────────────

class StatsResponse(BaseModel):
    """Response shape for /api/stats."""

    total_cves_processed: int = 0
    critical_count: int = 0
    high_count: int = 0
    medium_count: int = 0
    low_count: int = 0
    last_poll_time: Optional[datetime] = None
    next_poll_time: Optional[datetime] = None
    subscribers_count: int = 0
    kev_catalog_size: int = 0


class PollResponse(BaseModel):
    """Response shape for /api/poll."""

    status: str = "ok"
    new_cves_found: int = 0
    cves_processed: int = 0
    alerts_sent: int = 0
    errors: list[str] = Field(default_factory=list)


# ── Community Intelligence (Phase 4) ─────────────────────────────────────────

class WildReport(BaseModel):
    """Request body for reporting a CVE seen in the wild."""
    cve_id: str
    context: Optional[str] = None  # optional free-text, max 200 chars


class WildReportSummary(BaseModel):
    """Aggregated wild exploitation reports for a CVE."""
    cve_id: str
    report_count: int = 0
    last_reported_at: Optional[str] = None


class ResearcherNote(BaseModel):
    """Request body for adding a researcher note."""
    cve_id: str
    note: str           # max 500 chars
    author_alias: Optional[str] = None  # optional display name


class ResearcherNoteResponse(BaseModel):
    """Response shape for researcher notes."""
    id: int
    cve_id: str
    note: str
    author_alias: Optional[str] = None
    created_at: str


# ── Exploit Intelligence (Phase 4.5) ─────────────────────────────────────────

class ExploitEntry(BaseModel):
    """A single exploit entry from ExploitDB."""
    id: str
    description: str
    date: str
    type: str
    platform: str
    url: str


class PoCRepo(BaseModel):
    """A GitHub PoC repository with quality signals."""
    url: str
    stars: int = 0
    forks: int = 0
    last_updated: Optional[str] = None


class ExploitIntelligence(BaseModel):
    """
    Full exploit ecosystem intelligence for a single CVE.
    Stored separately from ProcessedCVE to avoid bloating the main table.
    """
    cve_id: str

    # Metasploit
    has_metasploit_module: bool = False
    metasploit_module_url: Optional[str] = None
    metasploit_module_path: Optional[str] = None

    # ExploitDB
    exploitdb_entries: list[ExploitEntry] = Field(default_factory=list)

    # GitHub PoC repos with quality signals
    poc_repos: list[PoCRepo] = Field(default_factory=list)

    # Exploit Maturity Score
    ems_score: int = 0
    ems_label: str = "RESEARCH"   # WEAPONIZED / FUNCTIONAL / THEORETICAL / RESEARCH

    # Computed commands
    metasploit_command: Optional[str] = None   # ready msfconsole command stub
    nuclei_command: Optional[str] = None       # ready nuclei command
    searchsploit_command: Optional[str] = None # searchsploit command

    # Meta
    last_updated: datetime = Field(default_factory=datetime.utcnow)


class ExploitIntelSummary(BaseModel):
    """Lightweight summary for feed cards — no full exploit details."""
    cve_id: str
    ems_score: int
    ems_label: str
    has_metasploit_module: bool
    has_nuclei_template: bool
    has_exploitdb_entry: bool
    poc_count: int
    max_poc_stars: int
    priority_score: int
    priority_label: str
    cvss_score: float
    description: str
    published: str


# ── Advisory Feed (Supply Chain Detection) ────────────────────────────────────

class AdvisoryRecord(BaseModel):
    """A security advisory from GitHub Advisory Database or OSV.dev."""

    ghsa_id: Optional[str] = None           # e.g., "GHSA-xxxx-xxxx-xxxx"
    cve_id: Optional[str] = None            # may be None for malware advisories
    source: str = "github"                  # "github" or "osv"
    ecosystem: str = "npm"                  # "npm", "PyPI", "Go", etc.
    severity: str = "unknown"               # "critical", "high", "medium", "low"
    summary: str = ""
    description: str = ""
    affected_packages: list[dict] = Field(default_factory=list)
    # Each dict: {"name": str, "ecosystem": str, "version_range": str}
    references: list[str] = Field(default_factory=list)
    published_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    is_malware: bool = False                # True for supply chain / malware advisories
    withdrawn: bool = False                 # True if advisory was withdrawn

File: dashboard/src/types/cve.ts
/* ── TypeScript types mirroring backend Pydantic models ─── */

export interface EnrichmentData {
  in_kev: boolean;
  epss_score: number;
  epss_percentile: number;
  poc_urls: string[];
  has_poc: boolean;
  greynoise_scanner_count: number;
  is_being_scanned: boolean;
  // Trend tracking
  previous_epss_score: number;
  previous_scanner_count: number;
  epss_trend: 'rising' | 'falling' | 'stable';
  scanner_trend: 'rising' | 'falling' | 'stable' | 'new';
  // Nuclei template (Phase 4)
  has_nuclei_template: boolean;
  nuclei_template_url?: string;
}

export interface AttackTechnique {
  technique_id: string;
  technique_name: string;
  tactic: string;
  tactic_phase: number;
  description: string;
  is_pivot: boolean;
}

export interface AIExplanation {
  summary: string;
  technical_detail: string;
  impact: string;
  remediation: string;
  tags: string[];
  affected_tech: string[];
  mitre_techniques: Array<{
    technique_id: string;
    technique_name: string;
    tactic: string;
    url: string;
  }>;
  // Phase 5.5 — AI Depth Upgrade fields
  vulnerability_class_analysis?: string;
  adversarial_context?: string;
  exploit_narrative?: string;
  attack_techniques?: AttackTechnique[];
}

export interface CVEReference {
  url: string;
  source?: string;
}

export interface ProcessedCVE {
  cve_id: string;
  description: string;
  published_date: string | null;
  last_modified: string | null;
  cvss_score: number;
  cvss_vector: string;
  cvss_version: string;
  enrichment: EnrichmentData;
  priority_score: number;
  priority_label: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  ai_explanation?: AIExplanation | null;
  references: string[];
  weaknesses: string[];
  processed_at: string;
  // Dynamic scoring & categories
  categories: string[];
  dynamic_score: number;
  last_rescored_at?: string;
}

export interface StatsResponse {
  total_cves_processed: number;
  critical_count: number;
  high_count: number;
  medium_count: number;
  low_count: number;
  last_poll_time: string | null;
  next_poll_time: string | null;
  subscribers_count: number;
  kev_catalog_size: number;
}

export interface PollResponse {
  status: string;
  new_cves_found: number;
  cves_processed: number;
  alerts_sent: number;
  errors: string[];
}

export interface HealthResponse {
  status: string;
  service: string;
  poller_active: boolean;
  telegram_active: boolean;
}

export type CVECategory =
  | 'ACTIVELY_EXPLOITED'
  | 'TRENDING'
  | 'JUST_DROPPED'
  | 'HIGH_EXPLOITABILITY'
  | 'NO_AUTH_REQUIRED';

export type PriorityLabel = 'ALL' | 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';

export type CategoryLabel = 'ALL' | CVECategory;

export type FilterState = {
  priority: PriorityLabel;
  category: CategoryLabel;
  kev_only: boolean;
  has_poc: boolean;
  search: string;
};

// Phase 4 — Community Intelligence

export interface WildReportSummary {
  cve_id: string;
  report_count: number;
  last_reported_at?: string;
}

export interface ResearcherNoteItem {
  id: number;
  cve_id: string;
  note: string;
  author_alias?: string;
  created_at: string;
}

export interface NucleiInfo {
  cve_id: string;
  has_template: boolean;
  template_url?: string;
  nuclei_command?: string;
}

// Phase 4.5 — Exploit Intelligence

export interface ExploitEntry {
  id: string;
  description: string;
  date: string;
  type: string;
  platform: string;
  url: string;
}

export interface PoCRepo {
  url: string;
  stars: number;
  forks: number;
  last_updated?: string;
}

export interface ExploitIntelligence {
  cve_id: string;
  has_metasploit_module: boolean;
  metasploit_module_url?: string;
  metasploit_module_path?: string;
  exploitdb_entries: ExploitEntry[];
  poc_repos: PoCRepo[];
  ems_score: number;
  ems_label: 'WEAPONIZED' | 'FUNCTIONAL' | 'THEORETICAL' | 'RESEARCH';
  metasploit_command?: string;
  nuclei_command?: string;
  searchsploit_command?: string;
  last_updated: string;
}

export interface ExploitIntelSummary {
  cve_id: string;
  ems_score: number;
  ems_label: string;
  has_metasploit_module: boolean;
  has_nuclei_template: boolean;
  has_exploitdb_entry: boolean;
  poc_count: number;
  max_poc_stars: number;
  priority_score: number;
  priority_label: string;
  cvss_score: number;
  description: string;
  published: string;
}

// ── Phase 5: Threat Intelligence Types ──────────────────────────────────────

export interface ThreatActor {
  id?: number;
  slug: string;
  name: string;
  aliases: string[];
  origin_country?: string;
  motivation: string;
  sophistication: string;
  description: string;
  targeted_sectors: string[];
  targeted_countries: string[];
  mitre_group_id?: string;
  mitre_url?: string;
  first_seen?: string;
  last_active?: string;
  is_active: boolean;
  cve_count?: number;
}

export interface ThreatActorDetail extends ThreatActor {
  exploited_cves: Array<{
    cve_id: string;
    confirmed: boolean;
    notes?: string;
    source_url?: string;
  }>;
}

export interface RansomwareCampaign {
  id?: number;
  actor_slug: string;
  actor_name?: string;
  campaign_name: string;
  cve_ids: string[];
  cves?: string[]; // alias for cve_ids in matrix view
  sectors: string[];
  countries: string[];
  status: "active" | "recent" | "historical";
  description: string;
  motivation?: string;
  origin_country?: string;
  source_url?: string;
}

export interface IOCResult {
  indicator: string;
  ioc_type: "ip" | "domain" | "url" | "hash";
  verdict: "malicious" | "suspicious" | "clean" | "unknown";
  risk_score: number;
  sources: {
    threatfox?: { hit: boolean; malware_family?: string; tags?: string[] };
    abuseipdb?: {
      confidence: number;
      reports: number;
      country?: string;
      isp?: string;
    };
    urlhaus?: { status?: string; tags?: string[]; urls_count?: number };
    greynoise?: {
      noise: boolean;
      riot: boolean;
      classification?: string;
      name?: string;
    };
  };
  related_cves: string[];
  cached: boolean;
}

export interface IOCFeedEntry {
  indicator: string;
  ioc_type: string;
  malware_family: string;
  tags: string[];
  threat_type: string;
  reported_at: string;
}

export interface SecurityNewsItem {
  id: number;
  title: string;
  url: string;
  source: string;
  published_at: string;
  summary?: string;
  mentioned_cves: string[];
  mentioned_actors: string[];
  tags: string[];
}

export interface BreachRecord {
  id: number;
  company_name: string;
  breach_date?: string;
  disclosed_date?: string;
  actor_slug?: string;
  actor_name?: string;
  cve_ids: string[];
  data_exposed: string[];
  records_count?: number;
  sectors: string[];
  description: string;
  source_urls: string[];
  verified: boolean;
}

export interface CVEContext {
  threat_actors: ThreatActor[];
  ransomware_groups: RansomwareCampaign[];
  news_articles: SecurityNewsItem[];
  breaches: BreachRecord[];
}

export interface FullCVEContext {
  cve: ProcessedCVE;
  context: CVEContext;
}

// ── Phase 7: Watchlist & Exposure Types ─────────────────────────────────────

export interface WatchlistItem {
  id: string;
  user_id: string;
  cpe_string: string;
  display_name: string;
  criticality: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';
  created_at: string;
}

export interface ExposureScore {
  score: number;
  critical_count: number;
  high_count: number;
  actively_exploited_count: number;
  top_cves: Array<{
    cve_id: string;
    priority_score: number;
    priority_label: string;
    description: string;
    in_kev: boolean;
    ai_summary: string;
  }>;
  calculated_at: string;
}

// ── Phase 8: Org, Asset, Triage, MSSP Types ─────────────────────────────────

export type OrgType = 'team' | 'mssp';
export type OrgRole = 'owner' | 'admin' | 'member' | 'viewer';
export type TriageStatus = 'new' | 'investigating' | 'remediation_planned' | 'mitigated' | 'wont_fix';

export interface Organization {
  id: string;
  name: string;
  org_type: OrgType;
  owner_id: string;
  created_at: string;
  updated_at: string;
  // Enriched by backend
  user_role?: OrgRole;
  members?: OrgMember[];
  member_count?: number;
  asset_count?: number;
  plan_limits?: { assets: number; members: number };
}

export interface OrgMember {
  id: string;
  org_id: string;
  user_id: string;
  member_role: OrgRole;
  joined_at: string;
  // Enriched from user_profiles
  email?: string;
  display_name?: string;
  avatar_url?: string;
}

export interface OrgInvite {
  id: string;
  org_id: string;
  email: string;
  role: OrgRole;
  token: string;
  expires_at: string;
  accepted: boolean;
  created_at: string;
}

export interface Asset {
  id: string;
  org_id: string;
  display_name: string;
  cpe_string: string;
  criticality: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';
  owner_name: string;
  notes: string;
  client_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface TriageItem {
  id: string;
  org_id: string;
  cve_id: string;
  client_id: string | null;
  status: TriageStatus;
  assignee_id: string | null;
  notes: string;
  sla_hours: number;
  sla_started_at: string | null;
  sla_due_at: string | null;
  created_at: string;
  updated_at: string;
  // Enriched by backend
  cve_data?: {
    description: string;
    priority_score: number;
    priority_label: string;
    cvss_score: number;
    in_kev: boolean;
    ai_summary: string;
  } | null;
  is_overdue?: boolean;
  assignee_name?: string;
  assignee_avatar?: string;
  matched_assets?: string[];
}

export interface TriageActivity {
  id: string;
  triage_item_id: string;
  user_id: string | null;
  action: string;
  detail: string;
  created_at: string;
}

export interface SLAConfig {
  priority: string;
  sla_hours: number;
  org_id?: string;
}

export interface OrgClient {
  id: string;
  org_id: string;
  name: string;
  contact_name: string;
  contact_email: string;
  created_at: string;
  updated_at: string;
}

export interface OrgExposureScore {
  score: number;
  critical_count: number;
  high_count: number;
  actively_exploited_count: number;
  top_cves: Array<{
    cve_id: string;
    priority_score: number;
    priority_label: string;
    description: string;
    in_kev: boolean;
    matched_assets?: string[];
  }>;
  calculated_at: string;
  client_id?: string | null;
}

export interface ComplianceSnapshot {
  cves: Array<{
    cve_id: string;
    severity: string;
    in_kev: boolean;
    first_detected: string;
    triage_status: TriageStatus;
    days_to_remediate: number | null;
    matched_assets: string[];
  }>;
  stats: {
    total_cves: number;
    critical_high_count: number;
    kev_count: number;
    mitigated_count: number;
    remediation_rate: number;
    sla_compliance_rate: number;
    avg_remediation_days: number;
  };
  date_range: { days: number; from: string };
}

export interface ClientSummary {
  client_id: string;
  exposure_score: number;
  open_triage: number;
  overdue_count: number;
}

